source/common/router/config_impl.cc

```diff
@@ -383,6 +383,38 @@ RouteConstSharedPtr PathRouteEntryImpl::matches(const Http::HeaderMap& headers,
   return nullptr;
 }
 
+RegexRouteEntryImpl::RegexRouteEntryImpl(const VirtualHostImpl& vhost,
+                                         const envoy::api::v2::Route& route,
+                                         Runtime::Loader& loader)
+    : RouteEntryImplBase(vhost, route, loader),
+      regex_(std::regex{route.match().regex(), std::regex::optimize}) {}
+
+void RegexRouteEntryImpl::finalizeRequestHeaders(Http::HeaderMap& headers) const {
+  RouteEntryImplBase::finalizeRequestHeaders(headers);
+
+  const Http::HeaderString& path = headers.Path()->value();
+  const char* query_string_start = findQueryStringStart(path);
+  ASSERT(std::regex_match(path.c_str(), query_string_start, regex_));
+  std::string matched_path(path.c_str(), query_string_start);
+  finalizePathHeader(headers, matched_path);
```

Comment:
Didn't we say that the entire path will be replaced? So, why even bother to do a regex match here?

Reply:
Okay, reading the implementation below, can we cache the matched path from matches() function, so that we avoid the additional regex match for every request.

Reply:
Though the route entry is readonly. :(. So, I guess caching is out of the picture